src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesCodeFixProvider.cs

```diff
@@ -0,0 +1,702 @@
+        where TBlockSyntax : TStatementSyntax
+        where TExpressionStatementSyntax : TStatementSyntax
+        where TLocalDeclarationStatementSyntax : TStatementSyntax
+        where TVariableDeclaratorSyntax : SyntaxNode
+        where TForEachStatementSyntax: TStatementSyntax
+        where TSwitchCaseBlockSyntax: SyntaxNode
+        where TSwitchCaseLabelOrClauseSyntax: SyntaxNode
+    {
+        private static readonly SyntaxAnnotation s_memberAnnotation = new SyntaxAnnotation();
+        private static readonly SyntaxAnnotation s_newLocalDeclarationStatementAnnotation = new SyntaxAnnotation();
+        private static readonly SyntaxAnnotation s_unusedLocalDeclarationAnnotation = new SyntaxAnnotation();
+
+        public sealed override ImmutableArray<string> FixableDiagnosticIds => ImmutableArray.Create(IDEDiagnosticIds.ExpressionValueIsUnusedDiagnosticId,
+                                                                                                    IDEDiagnosticIds.ValueAssignedIsUnusedDiagnosticId);
+
+        public sealed override Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            var diagnostic = context.Diagnostics[0];
+            var preference = AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.GetUnusedValuePreference(diagnostic);
+            var isRemovableAssignment = AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.GetIsRemovableAssignmentDiagnostic(diagnostic);
+
+            string title;
+            if (isRemovableAssignment)
+            {
+                // Recommend removing the redundant constant value assignment.
+                title = FeaturesResources.Remove_redundant_assignment;
+            }
+            else
+            {
+                // Recommend using discard/unused local for redundant non-constant assignment.
+                switch (preference)
+                {
+                    case UnusedValuePreference.DiscardVariable:
+                        if (IsForEachIterationVariableDiagnostic(diagnostic, context.Document))
+                        {
+                            // Do not offer a fix to replace unused foreach iteration variable with discard.
+                            // User should probably replace it with a for loop based on the collection length.
+                            return Task.CompletedTask;
+                        }
+
+                        title = FeaturesResources.Use_discard_underscore;
+                        break;
+
+                    case UnusedValuePreference.UnusedLocalVariable:
+                        title = FeaturesResources.Use_discarded_local;
+                        break;
+
+                    default:
+                        return Task.CompletedTask;
+                }
+            }
+
+            context.RegisterCodeFix(
+                new MyCodeAction(
+                    title,
+                    c => FixAsync(context.Document, diagnostic, context.CancellationToken),
```

Comment:
passing context.CancellationToken here will cause AggregateException errors as it is the wrong cancellation token when the fix is applied.  pass along 'c' here.